File: web/links.py
"""
   Link Ripper

   Wraps the Internet Archive CDX API to pull a list of archived
   snapshots for one or more URLs, and format them as plain text,
   CSV, or HTML links.
"""
import requests


class LinksError(Exception):
    """Raised when a CDX lookup fails outright (timeout, bad status, etc)."""

    def __init__(self, message, url=None):
        super().__init__(message)
        self.url = url


class Links():
    BASE_URL = "https://web.archive.org/cdx/search/cdx?url="
    DUPES = "&showDupeCount="
    COLLAPSE = "&collapse="

    # Front-end sends friendly names; CDX API wants timestamp:N granularity.
    # timestamp:8 - every day, timestamp:7 - every 10 days, timestamp:6 - every month
    TIMESTAMP_MAP = {
        "1_day": "timestamp:8",
        "10_day": "timestamp:7",
        "1_month": "timestamp:6",
        # allow passing the raw collapse value straight through too
        "timestamp:8": "timestamp:8",
        "timestamp:7": "timestamp:7",
        "timestamp:6": "timestamp:6",
    }

    HEADERS = {
        "User-Agent": "Link-Ripper/1.0 (+https://example.com; contact admin)"
    }

    def __init__(self):
        pass

    def _build_query_url(self, url, show_dupes="false", timestamp=""):
        """
            Build the CDX query URL for a single target url.
            Kept as a local value (not self.BASE_URL) so repeated calls
            on the same instance never contaminate each other.
        """
        query_url = self.BASE_URL + url
        query_url += self.DUPES + str(show_dupes).lower()

        if timestamp:
            collapse_value = self.TIMESTAMP_MAP.get(timestamp)
            if collapse_value is None:
                raise LinksError(
                    "Unknown timestamp option: {}".format(timestamp), url=url
                )
            query_url += self.COLLAPSE + collapse_value

        return query_url

    def _fetch_cdx(self, url, show_dupes="false", timestamp=""):
        """
            Hit the CDX API for a single url and return the raw response text.
            Raises LinksError on timeout, connection failure, or non-200 status.
        """
        query_url = self._build_query_url(url, show_dupes, timestamp)

        try:
            r = requests.get(query_url, timeout=300, headers=self.HEADERS)
        except requests.Timeout:
            raise LinksError("Timed out contacting archive.org for {}".format(url), url=url)
        except requests.RequestException as exc:
            raise LinksError("Request failed for {}: {}".format(url, exc), url=url)

        if r.status_code != 200:
            raise LinksError(
                "Archive.org returned HTTP {} for {}".format(r.status_code, url),
                url=url,
            )

        return r.text

    def _parse_links(self, text):
        """
            Shared parser: turn raw CDX text into a list of wayback URLs.
            This is the single source of truth for how a link string is
            built, so txt/csv/html all render the exact same URLs -
            they just wrap them differently.
        """
        links = []
        for line in text.splitlines():
            line = line.strip()
            if not line:
                continue
            split_line = line.split(" ")
            if len(split_line) < 3:
                # malformed/unexpected CDX line - skip rather than crash
                continue
            links.append(
                "http://wayback.archive.org/web/{}/{}".format(
                    split_line[1], split_line[2]
                )
            )
        return links

    def rip_links_url(self, url, show_dupes="false", timestamp="", format="txt"):
        """
            Extract links for a single url and return them already
            formatted as a string in the requested format.
            Raises LinksError on failure; raises ValueError for an
            unknown format; returns None (caller should treat as
            "no results") if the CDX API returned nothing.
        """
        text = self._fetch_cdx(url, show_dupes, timestamp)
        links = self._parse_links(text)

        if not links:
            return None

        return self._format_links(links, format)

    def rip_links_urls(self, urllist, show_dupes="false", timestamp="", format="txt"):
        """
            Handle a list of urls. Results for every url are merged into
            a single formatted string (no per-url separator/header line),
            in the same order the urls were given.

            Returns (formatted_string_or_None, list_of_errors).
            formatted_string_or_None is None if there were zero total
            links across every url. errors is a list of LinksError for
            any urls that failed - callers should still show the
            successful results and report which urls failed.
        """
        all_links = []
        errors = []

        for url in urllist:
            url = url.strip()
            if not url:
                continue
            try:
                text = self._fetch_cdx(url, show_dupes, timestamp)
            except LinksError as exc:
                errors.append(exc)
                continue

            links = self._parse_links(text)
            all_links.extend(links)

        if not all_links:
            return None, errors

        return self._format_links(all_links, format), errors

    def _format_links(self, links, format):
        if format == "txt":
            return self.format_txt(links)
        elif format == "csv":
            return self.format_csv(links)
        elif format == "html":
            return self.format_html(links)
        else:
            raise ValueError("Unknown format: {}".format(format))

    def format_txt(self, links):
        """
            Format a list of wayback links as plain text, one per line.
        """
        return "\n".join(links)

    def format_csv(self, links):
        """
            Format a list of wayback links as CSV (single column, one per row).
        """
        return "\n".join(links)

    def format_html(self, links):
        """
            Format a list of wayback links as HTML anchors:
            <a href="url">url</a>, one per line.
        """
        return "\n".join(
            '<a href="{0}">{0}</a>'.format(link) for link in links
        )

File: addon/lib/links.js
/**
 * Core Link Ripper logic, ported from links.py.
 *
 * Same shared design as the Flask version: buildQueryUrl() builds one CDX
 * request per URL, parseLinks() is the single source of truth for turning
 * raw CDX text into wayback URLs, and formatTxt/formatCsv/formatHtml all
 * wrap that same list - they never re-derive the links themselves.
 */

export const BASE_URL = "https://web.archive.org/cdx/search/cdx?url=";

// Friendly front-end names -> CDX API's collapse=timestamp:N granularity.
// timestamp:8 - every day, timestamp:7 - every 10 days, timestamp:6 - every month
export const TIMESTAMP_MAP = {
  "1_day": "timestamp:8",
  "10_day": "timestamp:7",
  "1_month": "timestamp:6",
  "timestamp:8": "timestamp:8",
  "timestamp:7": "timestamp:7",
  "timestamp:6": "timestamp:6",
};

export class LinksError extends Error {
  constructor(message, url) {
    super(message);
    this.name = "LinksError";
    this.url = url;
  }
}

/**
 * Build the CDX query URL for a single target url. Always builds a fresh
 * string - never mutates shared state - so repeated calls can't leak into
 * each other the way the original Python bug did.
 */
export function buildQueryUrl(url, showDupes = "false", timestamp = "") {
  let queryUrl = BASE_URL + url;
  queryUrl += "&showDupeCount=" + String(showDupes).toLowerCase();

  if (timestamp) {
    const collapseValue = TIMESTAMP_MAP[timestamp];
    if (!collapseValue) {
      throw new LinksError(`Unknown timestamp option: ${timestamp}`, url);
    }
    queryUrl += "&collapse=" + collapseValue;
  }

  return queryUrl;
}

/**
 * Fetch raw CDX text for a single url. Throws LinksError on network
 * failure or a non-200 response.
 */
export async function fetchCdx(url, showDupes = "false", timestamp = "") {
  const queryUrl = buildQueryUrl(url, showDupes, timestamp);

  let response;
  try {
    response = await fetch(queryUrl, {
      headers: { "User-Agent": "Link-Ripper-Addon/1.0" },
    });
  } catch (err) {
    throw new LinksError(`Request failed for ${url}: ${err.message}`, url);
  }

  if (!response.ok) {
    throw new LinksError(
      `Archive.org returned HTTP ${response.status} for ${url}`,
      url
    );
  }

  return response.text();
}

/**
 * Shared parser: raw CDX text -> list of wayback URLs.
 */
export function parseLinks(text) {
  const links = [];
  for (const rawLine of text.split("\n")) {
    const line = rawLine.trim();
    if (!line) continue;
    const parts = line.split(" ");
    if (parts.length < 3) continue; // malformed/unexpected line - skip
    links.push(`http://wayback.archive.org/web/${parts[1]}/${parts[2]}`);
  }
  return links;
}

export function formatTxt(links) {
  return links.join("\n");
}

export function formatCsv(links) {
  return links.join("\n");
}

export function formatHtml(links) {
  return links.map((link) => `<a href="${link}">${link}</a>`).join("\n");
}

export function formatLinks(links, format) {
  switch (format) {
    case "txt":
      return formatTxt(links);
    case "csv":
      return formatCsv(links);
    case "html":
      return formatHtml(links);
    default:
      throw new Error(`Unknown format: ${format}`);
  }
}

/**
 * Fetch + merge results for a list of urls. Links from every url are
 * concatenated in order with no per-url separator/header line, matching
 * the Flask app's behavior. Returns { result, errors }: result is null
 * if there were zero total links across every url; errors lists any
 * urls that failed so callers can still show the successful results.
 */
export async function ripLinksUrls(
  urlList,
  showDupes = "false",
  timestamp = "",
  format = "txt"
) {
  const allLinks = [];
  const errors = [];

  for (const rawUrl of urlList) {
    const url = rawUrl.trim();
    if (!url) continue;

    try {
      const text = await fetchCdx(url, showDupes, timestamp);
      allLinks.push(...parseLinks(text));
    } catch (err) {
      if (err instanceof LinksError) {
        errors.push(err);
      } else {
        errors.push(new LinksError(err.message, url));
      }
    }
  }

  if (allLinks.length === 0) {
    return { result: null, errors };
  }

  return { result: formatLinks(allLinks, format), errors };
}
